### PR TITLE
Analyze all load/store users of struct fields

### PR DESCRIPTION
**Pull request by tcwanggoogle**
This CL performs analysis to find out all field accesses in the IR, from the definitions of globals and local structs, or pointers to structs. The results can be used to build CPG in the analysis. It alone can be used on Chrome to get the problem size and get some stats for further analysis. **It forks from and removes parts of the IR branch to reduce workload on reviewing!**

### CODE REVIEW

File: include/llvm/Transforms/IPO/StructFieldCacheAnalysis.h [hunk 1]
```diff
@@ -1,18 +1,18 @@
-// llvm/Tranforms/IPO/StructFieldCacheAnalysis.h - Performs Cache-Aware Structure Analysis- C++ -*-===//
```
Comment (eramandev) [inline include/llvm/Transforms/IPO/StructFieldCacheAnalysis.h:1]: Are these intentional changes?

File: lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 2]
```diff
@@ -573,6 +573,12 @@ void SelectionDAG::RemoveDeadNodes(SmallVectorImpl<SDNode *> &DeadNodes) {
   while (!DeadNodes.empty()) {
     SDNode *N = DeadNodes.pop_back_val();
```
Comment (eramandev) [inline lib/CodeGen/SelectionDAG/SelectionDAG.cpp:575]: Why is this needed?

File: lib/Transforms/IPO/PassManagerBuilder.cpp [hunk 3]
```diff
@@ -911,6 +919,9 @@ void PassManagerBuilder::populateLTOPassManager(legacy::PassManagerBase &PM) {
     //TODO: Need to make sure the profile-use pass has been executed
     PM.add(createStructFieldCacheAnalysisPass());
   }
+
```
Comment (eramandev) [inline lib/Transforms/IPO/PassManagerBuilder.cpp:914]: Your pass is not doing any transforms, so it doesn't make sense to run the verifier especially since it has already been run before yours.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 4]
```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+class StructFieldAccessInfo;
+class StructFieldAccessManager
+{
+  /* This class is used to keep track of all StructFieldAccessInfo objects in the program
+     and make sure only one StructFieldAccessInfo object for each type of struct declared
+     in the program.
+  */
+ public:
+  enum StructDefinitionType{
+    GlobalStruct,
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: You could prefix the enum entries with SDT_ (for StructDefinitionType). Same for the enum below.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 5]
```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+INITIALIZE_PASS_DEPENDENCY(BlockFrequencyInfoWrapperPass)
 INITIALIZE_PASS_END(StructFieldCacheAnalysisPass, "struct-field-cache-analysis", "Struct Field Cache Analysis", false, false)
 ModulePass *llvm::createStructFieldCacheAnalysisPass() { return new StructFieldCacheAnalysisPass; }
 
-static bool performStructFieldCacheAnalysis(Module &M)
+namespace llvm{
+typedef unsigned FieldNumType;
+
+class StructFieldAccessInfo;
+class StructFieldAccessManager
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: It is better to doxygen comments for the data structures and functions you have to keep this in sync with upstream coding style guidelines.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 6]
```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+  enum StructDefinitionType{
+    GlobalStruct,
+    GlobalStructPtr,
+    GlobalStructPtrPtr,
+    LocalStruct,
+    LocalStructPtr,
+    LocalStructPtrPtr
+  };
+
+  StructFieldAccessManager(const Module& M, function_ref<BlockFrequencyInfo *(Function &)> L):
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Use a more descriptive name for the second argument.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 7]
```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+      CurrentModule(M), LookupBFI(L), StatCounts(Stats::MaxNumStats) {};
+  // Check if the struct type is created before; if not, create a new StructFieldAccessInfo object for it
+  StructFieldAccessInfo* createOrGetStructFieldAccessInfo(const Type* T, const StructDefinitionType ST);
+  // Retrieve the pointer to the previous created StructFieldAccessInfo object for the type
+  StructFieldAccessInfo* getStructFieldAccessInfo(const Type* T) const;
+  // Retrieve execution count for a basic block
+  Optional<uint64_t> getExecutionCount(const BasicBlock* BB) const{
+    Function* func = const_cast<Function*>(BB->getParent());
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Variable names have to start with an uppercase letter.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 8]
```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+    essential information for cache-aware struct field analysis.
+  */
+ public:
+  StructFieldAccessInfo(const Type* T, const StructFieldAccessManager::StructDefinitionType ST, const Module& MD, const StructFieldAccessManager* M, const DICompositeType* D):
+      Eligiblity(true),
+      CurrentModule(MD),
+      StructureType(dyn_cast<StructType>(T)),
+      StructDefinition(ST),
+      DebugInfo(D),
+      NumElements(StructureType->getNumElements()),
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: You're dyncasting T into StructureType, which means StructureType could be nullptr, so you should not call getNumElements on it.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 9]
```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+      StructManager(M),
+      StatCounts(StructFieldAccessManager::Stats::MaxNumStats) {}
+
+  ~StructFieldAccessInfo() {}
+
+  StructFieldAccessManager::StructDefinitionType getStructDefinition() const { return StructDefinition; }
+  bool isEligible() const { return Eligiblity; }
+  // Analyze a value pointing to a struct and collect struct access from it. It can be allocas/function args/globals
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: I prefer leaving a blank line before comments associated with declarations to improve readability.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 10]
```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+    return StructManager->getExecutionCount(I->getParent());
+  }
+  // Print all instructions that access any struct field
+  void debugPrintAllStructAccesses(raw_ostream& OS);
+
+  // For stats
+  void addStats(unsigned Category, unsigned Opcode = 0) {
+    StatCounts[Category]++;
+    if (Category == StructFieldAccessManager::Stats::GepUnknownUse){
+      UnknownOpcodes.insert(Opcode);
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Why not have a map from unknown op code to count to understand their distribution?

File: include/llvm/Transforms/IPO/StructFieldCacheAnalysis.h (hunk 1 above)
Comment (tcwanggoogle) [inline include/llvm/Transforms/IPO/StructFieldCacheAnalysis.h]: Yes. I add the "*-" because I thought it's useless before but it looks like I should put it there according to LLVM programming guide

Comment (tcwanggoogle) [reply include/llvm/Transforms/IPO/StructFieldCacheAnalysis.h:1]: Yes. I add the "*-" because I thought it's useless before but it looks like I should put it there according to LLVM programming guide

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 11]
```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+  virtual void emitInstructionAnnot(const Instruction *I,
+                                    formatted_raw_ostream &OS) {
+    if (StructManager == NULL)
+      return;
+    if (auto pair = StructManager->getFieldAccessOnInstruction(I)){
+      OS.changeColor(raw_ostream::GREEN, false, false);
+      auto* type = pair.getValue().first;
+      assert(isa<StructType>(type));
+      if (dyn_cast<StructType>(type)->isLiteral())
+        OS << "; [Instruction is memory access on field " << pair.getValue().second << " of a literal struct.] ";
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Suggestion: you could shorten the message to remove clutter.

**Review by eramandev — COMMENTED** [review]
Some initial comments - mostly stylistic or nits. You should also write test case(s).

File: lib/CodeGen/SelectionDAG/SelectionDAG.cpp (hunk 2 above)
Comment (tcwanggoogle) [reply lib/CodeGen/SelectionDAG/SelectionDAG.cpp:575]: This is needed to fix an assertion error in Chrome. Caroline found this could fix that problem. https://reviews.llvm.org/D33731

File: lib/Transforms/IPO/PassManagerBuilder.cpp (hunk 3 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/PassManagerBuilder.cpp:914]: Thanks! I think it's added when GitHub merging two different versions. I didn't intend to do so...

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 5 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Fixed.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 4 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Is there any requirement on the name of enum? e.g. Do I have to end with "Kind" or "Type"? I checked that LLVM says it needs to be "Kind" as a discriminator for a union, or an indicator of a subclass. But I don't use it for the case.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 6 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Fixed. I also used a better name is later branches when more functions are needed...

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 7 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Missed in this part. Fixed.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 8 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: I checked type of T before calling constructor. I'll move the assignment of StructureType and NumElement into the constructor body to make sure they can be assigned.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 9 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Fixed. Thanks!

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 11 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: I try to remove some unnecessary words to shorten this message.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 10 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: I wasn't expect to make this opcode stats complicated. I'll change it to a map.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 12]
```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+
+  // Find all structs declared by allocas
+  for (auto &F : M){
+    if (F.isDeclaration())
+      continue;
+    //Find all alloca of structs inside the function body
+    for (auto &BB : F){
+      for (auto &I: BB){
+        if (I.getOpcode() == Instruction::Alloca){
+          assert(I.getType()->isPointerTy());
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: You don't need to assert conditions that are true in LLVM IR.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 13]
```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+
+void StructFieldAccessInfo::analyzeUsersOfStructValue(const Value* V)
+{
+  assert(V->getType()->isPointerTy() && V->getType()->getPointerElementType()->isStructTy());
+  for (auto *U : V->users()){
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Analyzing user of " << *V << ": " << *U << "\n");
+    //assert(isa<Instruction>(U) || isa<Operator>(U));
+    if (isa<Instruction>(U)){
+      auto *Inst = dyn_cast<Instruction>(U);
+      if (Inst->getOpcode() != Instruction::GetElementPtr){
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Use isa instead of getOpcode

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 14]
```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+          if (F->isDeclaration()){
+            // If a struct is passed to a function not declared in the program, we can't analyze it...
+            Eligiblity = false;
+          }
+        }
+        continue;
+      }
+      addFieldAccessFromGEP(Inst);
+    }
+    else if (isa<Operator>(U)){
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: May be I'm missing something here. Don't you care only when the struct is used in an instruction? Can you remove this else if part?

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 15]
```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+  assert(V->getType()->isPointerTy() && V->getType()->getPointerElementType()->isStructTy());
+  for (auto *U : V->users()){
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Analyzing user of " << *V << ": " << *U << "\n");
+    //assert(isa<Instruction>(U) || isa<Operator>(U));
+    if (isa<Instruction>(U)){
+      auto *Inst = dyn_cast<Instruction>(U);
+      if (Inst->getOpcode() != Instruction::GetElementPtr){
+        // Only support access struct through GEP for now
+        if (Inst->getOpcode() == Instruction::Call){
+          auto* F = dyn_cast<CallInst>(Inst)->getCalledFunction();
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: If it is a call to a function whose definition is available, you don't seem to do anything here. If you want to handle this later, add TODOs.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 16]
```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+    if (isa<Instruction>(User)){
+      auto* Inst = dyn_cast<Instruction>(User);
+      if (Inst->getOpcode() == Instruction::Load)
+        addFieldAccessNum(Inst, FieldLoc);
+      else if (Inst->getOpcode() == Instruction::Store){
+        if (U == Inst->getOperand(1))
+          addFieldAccessNum(Inst, FieldLoc);
+      }
+      else{
+        if (Inst->getOpcode() == Instruction::Call || Inst->getOpcode() == Instruction::Invoke){
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Add comments saying you are ignoring anything other than load and store now.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 17]
```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+  }
+  return ret;
+}
+
+FieldNumType StructFieldAccessInfo::calculateFieldNumFromGEP(const User* U) const
+{
+  DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Calculating field number from GEP: " << *U << "\n");
+  //Operand 0 should be a pointer to the struct
+  assert(isa<GetElementPtrInst>(U) || isa<GEPOperator>(U));
+  auto* Op = U->getOperand(0);
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Use getPointerOperand instead

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 18]
```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+  // Make sure Operand 0 is a struct type and matches the current struct type of StructFieldAccessInfo
+  assert(Op->getType()->isPointerTy() && Op->getType()->getPointerElementType()->isStructTy() && Op->getType()->getPointerElementType() == StructureType);
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: The first part of the assert is not needed - you know that the type of the operand 0 of a GEP is a pointer type. In general add asserts only for invariants that are specific to your code.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 17 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: I didn't use that is because U can be either a GetElementPtrInst or a GEPOperator. It needs extra code to handle the two difference cases. Please let me know if I still need to change it.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 18 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Got it. I put it there when I first try it and still understanding each LLVM functions.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 16 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: I just removed this part in a commit before. I figured it could be a problem. I don't why it's not updated...

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 13 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Got it. I am just curious that is there any special reason not to use getOpcode()? Also, do I need to do the same thing for everywhere I check the opcode?

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 14 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Because I found GEP can also be an Operator (GEPOperator) that uses a struct field, not just GetElementPtrInst

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 12 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Again, it was the time when I am not sure about some of the behaviors... Thank you for catching it!

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp (hunk 4 above)
Comment (eramandev) [reply lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Kind is used in specific circumstances as you describe. Not applicable here.

File: lib/Transforms/IPO/StructFieldCacheAnalysis.cpp [hunk 19]
```diff
@@ -36,30 +46,657 @@ class StructFieldCacheAnalysisPass : public ModulePass {
...
+      CurrentModule(MD),
+      StructureType(NULL),
+      StructDefinition(ST),
+      DebugInfo(D),
+      NumElements(0),
+      StructManager(M),
+      StatCounts(StructFieldAccessManager::DebugStats::DS_MaxNumStats)
+  {
+    assert(T && T->isStructTy());
+    StructureType = dyn_cast<StructType>(T);
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysis.cpp]: Use cast instead of dyn_cast if you know it is always passed the right type. Actually, it is then preferable to cast into StructType at the caller of this constructor and make this take StructType *ST.

**Review by eramandev — APPROVED** [review]
The test needs to take in PGO info, but you can do it as a subsequent patch. Do a round of clang-format before submitting. You can also remove some obvious asserts, but again not critical now.
